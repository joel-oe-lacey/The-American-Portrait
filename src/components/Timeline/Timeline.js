import React, { Component } from 'react';
import './Timeline.scss';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

//conditionally render forward or back based on length 
//pieces array come in as props
//deconstruct to display certain info 

//link to art piece on 

export class Timeline extends Component {
    constructor({ collection }) {
        super()
        this.state = {
            collection,
            i: 0,
        }
    }

    componentDidMount() {
        //fetch
        //run through helper
        //add to store
        //add to local storage 
    }

    changePiece = (direction) => {
        let { i, collection } = this.state;
        let endIndex = collection.length - 1;
        let newIndex;

        if (direction === 'left') {
            newIndex = i ? i - 1 : 0;
        } else {
            newIndex = ((i === endIndex) ? i : i + 1);
        }

        this.setState({
            i: newIndex
        });
    }

    render() {
        const { collection, i } = this.state;
        let endIndex = collection.length - 1;
        //abstract index checking into another function to reduce redundancy 

        return (
        <section className="carousel">
            <section className="carousel-disp">
                <Link to={`piece/${collection[i].objectid}`} className='carousel-link' >
                    <img className='carousel-img' 
                    src={collection[i].primaryimageurl} 
                    alt={collection[i].description}  />
                </Link>
            </section>
            <section className="carousel-timeline">
                <button className='carousel-arrow-left' onClick={() => this.changePiece('left')}>←</button>
                    <h2 className="timeline-prev">{i ? collection[i-1].dateend : collection[i].dateend}</h2>
        <h1 className="timeline-curr">{collection[i].dateend}</h1>
                    <h2 className="timeline-next">{i === endIndex ? collection[i].dateend : collection[i + 1].dateend}</h2>
                <button className='carousel-arrow-right' onClick={() => this.changePiece('right')}>→</button>
            </section>
        </section>
    )}
}

export const mapStateToProps = state => ({
    regions: state.regions,
});

export default connect(mapStateToProps)(Timeline)
