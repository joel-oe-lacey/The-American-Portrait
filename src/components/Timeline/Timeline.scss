.carousel {
    height: 100%;
    width: 100%;

    background-color: #202020;
    color: white;
    display: flex;
    flex-direction: column;
    align-items: center;
}

.carousel-disp {
    height: 75%;
    width: 100%;
    display: grid;
    grid-template-columns: repeat(10, 10%);
    grid-template-areas:
    ". prevImg . currImg currImg currImg currImg . nextImg .";
    align-items: center;
    text-align: center;
    max-height: 100%;
    grid-template-rows: 100%;
}

.carousel-timeline {
    height: 15%;
    width: 100%;

    //might want to grid this for a more even display
    display: grid;
    grid-template-columns: repeat(10, 10%);
    grid-template-areas:
    ". left . prevTime currTime currTime nextTime . right .";
    align-items: center;
    text-align: center;
}

.carousel-link {
    height: 70%;
    width: 100%;
    grid-area: currImg;
}

.carousel-title {
    height: 10%;
    width: 50%;
    text-align: center;
    font-size: 1.2em;
}

.carousel-img-prev {
    height: 40%;
    width: 100%;
    object-fit: scale-down;
    grid-area: prevImg;
}

.carousel-img-next {
    height: 40%;
    width: 100%;
    object-fit: scale-down;
    grid-area: nextImg;
}

.carousel-img {
    height: 100%;
    width: 100%;
    object-fit: contain;
}

.timeline-prev {
    grid-area: prevTime;
    border-bottom: 1px solid white;
}

.timeline-next {
    grid-area: nextTime;
    border-bottom: 1px solid white;
}

.timeline-curr {
    grid-area: currTime;
    border: 1px solid white;
}

.carousel-arrow-left {
    grid-area: left;
}

.carousel-arrow-right {
    grid-area: right;
}
