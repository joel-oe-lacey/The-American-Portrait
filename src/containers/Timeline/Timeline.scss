@import '../../variables';

.carousel {
    height: 100%;
    width: 100%;

    background-color: #202020;
    color: white;
    display: flex;
    flex-direction: column;
    align-items: center;
}

.carousel-disp {
    height: 75%;
    width: 100%;
    display: grid;
    grid-template-columns: repeat(10, 10%);
    grid-template-areas:
    ". prevImg . currImg currImg currImg currImg . nextImg .";
    align-items: center;
    text-align: center;
    max-height: 75%;
    grid-template-rows: 100%;
}

.carousel-position {
    height: 5%;
    width: 75%;
    display: flex;
    flex-direction: row;
    justify-content: space-around;
}

.position-inactive {
    font-size: 1rem;
}

.position-active {
    font-size: 2rem;
}

.carousel-timeline {
    height: 10%;
    width: 100%;

    //might want to grid this for a more even display
    display: grid;
    grid-template-columns: repeat(10, 10%);
    grid-template-areas:
    ". left . prevTime currTime currTime nextTime . right .";
    align-items: center;
    text-align: center;
    margin: 1%;

    button {
        background: transparent;
        font-style: italic;
        color: white;
        height: 50%;
        @include roundedBorder(40%);
    }
}

.carousel-link {
    height: 70%;
    width: 100%;
    grid-area: currImg;
    object-fit: scale-down;
}

.carousel-title {
    height: 10%;
    width: 50%;
    text-align: center;
    font-size: 1.2rem;
    text-overflow: ellipsis;
    white-space: nowrap;
    overflow: hidden;
}

.carousel-img-prev {
    height: 40%;
    width: 100%;
    object-fit: scale-down;
    grid-area: prevImg;
}

.carousel-img-next {
    height: 40%;
    width: 100%;
    object-fit: scale-down;
    grid-area: nextImg;
}

.carousel-img {
    height: 100%;
    width: 100%;
    object-fit: scale-down;
}

.timeline-prev {
    grid-area: prevTime;
    @include textFlex();
    font-size: .8rem;
}

.timeline-next {
    grid-area: nextTime;
    @include textFlex();
    font-size: .8rem;
}

.timeline-curr {
    grid-area: currTime;
    border: 1px solid white;
    font-size: 1rem;
    height: 100%;
    @include roundedBorder(40%);
    @include textFlex();
}

.carousel-arrow-left {
    grid-area: left;
}

.carousel-arrow-right {
    grid-area: right;
}
