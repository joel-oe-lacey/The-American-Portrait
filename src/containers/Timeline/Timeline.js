import React, { Component } from 'react';
import './Timeline.scss';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import Invalid from '../Invalid/Invalid';
export class Timeline extends Component {
    constructor({ collection }) {
        super()
        this.state = {
            collection,
            i: 0,
        }
    }

    changePiece = (direction) => {
        let { i, collection } = this.state;
        let endIndex = collection.length - 1;
        let newIndex;

        if (direction === 'left') {
            newIndex = i ? i - 1 : 0;
        } else {
            newIndex = ((i === endIndex) ? i : i + 1);
        }

        this.setState({
            i: newIndex
        });
    }

    render() {
        const { i } = this.state;
        const { collection } = this.props;
        let endIndex = collection.length - 1;
        //abstract index checking into another function to reduce redundancy 
        //perhaps make below a component
        const positionIndicator = collection.map((item, mapIndex) => {
            if(i === mapIndex) {
                return <h4 key={mapIndex} className="position-active">.</h4>
            }
            return <h4 key={mapIndex}className="position-inactive">.</h4>
        })

        if (!collection.length) {
            return <Invalid />
        } else {
            return (
            <section className="carousel">
                <Link to={`/`} className='return' >
                    Home
                </Link>
                <section className="carousel-disp">
                        {(i !== 0) && <img className='carousel-img-prev'
                            src={collection[i-1].primaryimageurl}
                            alt={collection[i-1].description} />}
                    <Link to={`piece/${collection[i].objectid}`} className='carousel-link' >
                        <img className='carousel-img' 
                        src={collection[i].primaryimageurl} 
                        alt={collection[i].description}  />
                    </Link>
                        {(i !== endIndex) && <img className='carousel-img-next'
                            src={collection[i + 1].primaryimageurl}
                            alt={collection[i + 1].description} />}
                </section>
                <h1 className="carousel-title">{collection[i].title}</h1>
                <section className="carousel-position">
                    {positionIndicator}
                </section>
                <section className="carousel-timeline">
                        {(i !== 0) && <button className='carousel-arrow-left' onClick={() => this.changePiece('left')}>Backwards</button>}
                        {i ? <span className="timeline-prev">
                            <h2>{collection[i-1].dateend}</h2>
                            <h2>Previous</h2>
                        </span> : ''}
                        <span className="timeline-curr">
                            <h1>{collection[i].dateend}</h1>
                            <h2>Current</h2>
                        </span>
                        {i !== endIndex ?<span className="timeline-next">
                            <h2>{collection[i + 1].dateend}</h2>
                            <h2>Next</h2>
                        </span> : ''}
                        {(i !== endIndex) && <button className='carousel-arrow-right' onClick={() => this.changePiece('right')}>Forwards</button>}
                </section>
            </section>
        )}
    }
}

export const mapStateToProps = state => ({
    collection: state.collections,
});

export default connect(mapStateToProps)(Timeline)
