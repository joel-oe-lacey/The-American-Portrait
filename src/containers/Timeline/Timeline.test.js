import React from 'react';
import { shallow } from 'enzyme';
import { Timeline } from './Timeline';

//test map state to props 
//test indexing functionality 
//different snapshot for each conditional render? 

describe('Timeline', () => {
    let wrapper;

    beforeEach(() => {
    wrapper = shallow(<Timeline />);
    });

    it('Should equal snapshot', () => {
        const wrapper = shallow(<Timeline />);
        expect(wrapper).toMatchSnapshot();
    });

    it('Should subtract index for left click', () => {
        wrapper.instance().setState({i: 10, collection: [1,2,3,4,5]})

        wrapper.instance().changePiece('left');

        expect(wrapper.state('i')).toEqual(9);
    });

    it('Should add to image index for right click', () => {
        wrapper.instance().setState({ i: 10, collection: [1, 2, 3, 4, 5] })

        wrapper.instance().changePiece('right');

        expect(wrapper.state('i')).toEqual(11);
    });
});