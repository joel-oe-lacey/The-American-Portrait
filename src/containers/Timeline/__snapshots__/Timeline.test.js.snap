// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline Should equal snapshot 1`] = `
<section
  className="carousel"
>
  <section
    className="carousel-disp"
  >
    <Link
      className="carousel-link"
      to="piece/354191"
    >
      <img
        className="carousel-img"
        src="https://nrs.harvard.edu/urn-3:HUAM:761794"
      />
    </Link>
    <img
      className="carousel-img-next"
      src="https://nrs.harvard.edu/urn-3:HUAM:761793"
    />
  </section>
  <h1
    className="carousel-title"
  >
    House and [blue] Car, Near Akron, Alabama
  </h1>
  <section
    className="carousel-position"
  >
    <h4
      className="position-active"
      key="0"
    >
      .
    </h4>
    <h4
      className="position-inactive"
      key="1"
    >
      .
    </h4>
  </section>
  <section
    className="carousel-timeline"
  >
    <span
      className="timeline-curr"
    >
      <h1>
        1978
      </h1>
      <h2>
        Current
      </h2>
    </span>
    <span
      className="timeline-next"
    >
      <h2>
        1972
      </h2>
      <h2>
        Next
      </h2>
    </span>
    <button
      className="carousel-arrow-right"
      onClick={[Function]}
    >
      Forwards
    </button>
  </section>
</section>
`;
